examples/authResolvers.js

```diff
@@ -0,0 +1,84 @@
+const fs = require('fs')
+const path = require('path')
+const express = require('express')
+const session = require('express-session')
+const Keycloak = require('keycloak-connect')
+const { ApolloServer, gql } = require('apollo-server-express')
+
+const { KeycloakContextProvider, KeycloakTypeDefs, auth, hasRole } = require('../')
+
+const app = express()
+
+const memoryStore = new session.MemoryStore()
+
+const graphqlPath = '/graphql'
+
+
+app.use(session({
+  secret: 'secret',
+  resave: false,
+  saveUninitialized: true,
+  store: memoryStore
+}))
+
+const keycloakConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, './config/keycloak.json')))
+
+const keycloak = new Keycloak({
+  store: memoryStore
+}, keycloakConfig)
+
+// Install general keycloak middleware
+app.use(keycloak.middleware({
+  admin: graphqlPath
+}))
+
+// Protect the main route for all graphql services
+// Disable unauthenticated access
+app.use(graphqlPath, keycloak.middleware())
+
+const typeDefs = gql`
+  type Query {
+    greetings: [String]!
+  }
+
+  type Mutation {
+    addGreeting(greeting: String!): String!
+  }
+`
+
+const greetings = [
+  'hello world!'
+]
+
+const resolvers = {
+  Query: {
+    greetings: () => greetings
+  },
+  Mutation: {
+    addGreeting: auth(hasRole('developer')((obj, { greeting }, context, info) => {
+      greetings.push(greeting)
+      return greeting
+    }))
+  }
+}
+
+// Initialize the voyager server with our schema and context
+const options ={
+  typeDefs: [KeycloakTypeDefs, typeDefs],
+  resolvers,
+  context: ({ req }) => {
+    return {
+      auth: new KeycloakContextProvider({ req })
```

Comment:
Do we depend on auth as key? If yes we need.to gracefully fail if not there. If you use directive and have no auth it should fail badly. Same if keycloak middleware is not there. There may be issue with multiple express app objects noy sharing the same middleware. We need to make sure that any of this developer errors will not mean that users will get access to stuff

Reply:
You're 100% right. More work is needed to protect the developer from these kinds of errors, and to also help them understand when things aren't configured properly.